### ISSUE TITLE
Memory issues in multithread context

### ISSUE THREAD

**Issue by atlascoder** [aws/aws-sdk-cpp#832]
### What platform/OS are you using?
MacOS, Android, iOS

### What compiler are you using? what version?
Clang 3.8
Android GCC 4.9 toolchain
Qt 5.9.3 on MacOS 10.12.6

### What's your CMake arguments?

for Android:
cmake -DBUILD_ONLY="core;sts;iot;iam;identity-management;cognito-sync;cognito-idp;cognito-identity;access-management;logs" -DBUILD_SHARED_LIBS=OFF -DTARGET_ARCH=ANDROID -DENABLE_TESTING=OFF -DNDK_DIR=/Users/anton.titkov/Library/Android/android-ndk-r12b -DANDROID_STL=gnustl_shared -DCMAKE_BUILD_TYPE=Release  ../src/aws-sdk-cpp

for MacOS (development host):
cmake -DBUILD_ONLY="core;sts;iot;iam;identity-management;cognito-sync;cognito-idp;cognito-identity;access-management;logs" -DBUILD_SHARED_LIBS=OFF -DENABLE_TESTING=OFF -DCMAKE_BUILD_TYPE=Release ../src/aws-sdk-cpp/

for iOS (generating for building under Xcode):
cmake -DBUILD_ONLY="core;sts;iot;iam;identity-management;cognito-sync;cognito-idp;cognito-identity;access-management;logs" -DBUILD_SHARED_LIBS=OFF -DENABLE_TESTING=OFF -DCMAKE_BUILD_TYPE=Release -G Xcode ../src/aws-sdk-cpp/

### Can you provide a TRACE level log? (sanitize any sensitive information)

> 1  Aws::Http::CreateHttpRequest(Aws::Http::URI const&, Aws::Http::HttpMethod, std::function<std::basic_iostream<char, std::char_traits<char>> * ()> const&)                                      0x10012e768    
2  Aws::Client::AWSClient::AttemptOneRequest(Aws::Http::URI const&, Aws::AmazonWebServiceRequest const&, Aws::Http::HttpMethod, const char *) const                                              0x1000fcd3d    
3  Aws::Client::AWSClient::AttemptExhaustively(Aws::Http::URI const&, Aws::AmazonWebServiceRequest const&, Aws::Http::HttpMethod, const char *) const                                            0x1000fb805    
4  Aws::Client::AWSJsonClient::MakeRequest(Aws::Http::URI const&, Aws::AmazonWebServiceRequest const&, Aws::Http::HttpMethod, const char *) const                                                0x1000ff94d    
5  Aws::CognitoSync::CognitoSyncClient::UpdateRecords(Aws::CognitoSync::Model::UpdateRecordsRequest const&) const                                                                                0x100376891    
6  CognitoSyncCommand::sendCommand(QString const&, QString const&, int)                                                                                     CognitoSyncCommand.cpp           103 0x1000c4148    
7  ControllerAPI::sendAwsCommand(QString const&, int)                                                                                                       ControllerAPI.cpp                263 0x1000c0993    
8  QtConcurrent::VoidStoredMemberFunctionPointerCall2<void, ControllerAPI, QString const&, QString, int, int>::runFunctor()                                 qtconcurrentstoredfunctioncall.h 929 0x1000c310c    
9  QtConcurrent::RunFunctionTask<void>::run()                                                                                                               qtconcurrentrunbase.h            136 0x1000395e1    
10 non-virtual thunk to QtConcurrent::RunFunctionTask<void>::run()                                                                                          qtconcurrentrunbase.h                0x100039769    
11 QThreadPoolThread::run()                                                                                                                                                                      0x101908c2e    
12 QThreadPrivate::start(void *)                                                                                                                                                                 0x10190bf6f    
13 _pthread_body                                                                                                                                                                                 0x7fffb053e93b 
14 _pthread_start                                                                                                                                                                                0x7fffb053e887 
15 thread_start                                                                                                                                                                                  0x7fffb053e08d 

*****

I have this issue on all platforms and it means that maust be I am using AWS lib not in right way. The error occurs always when I close the app while some AWS client is still creating (constructor is not completed). In this example the error occurs with CognitSyncClient, but it occurs on other Clients (at least CognitoIdentityClient), derived from AWSJsonClient (or even Aws::Client).

Clients are executed in separate threads under QtConcurrent - a kind of ThreadPool manager in Qt.
I also try to wait for thread stops before Aws::ShutdownApi and exit.

How to do right in this case?

**Comment by singku** [comment]
>I also try to wait for thread stops before Aws::ShutdownApi and exit.

This is actually the right way. But make sure every thread stops and exits before shutdown. It's better to embrace any AWS related code in braces like:
```
Aws::InitAPI
{
   //your code
}
Aws::ShutdownAPI
```

Another question is do you have customized memory management? If you do, make sure you don't have undestructed global variables before calling Aws::ShutdownAPI

**Comment by atlascoder** [comment]
Thanks @singku , maybe my threads are really not completed before Aws::ShutdownAPI - I will check.
I dont use customized memory management, and could you point to me some links on what is it needed for?

**Comment by JonathanHenson** [comment]
Custom memory management is mainly for video game customers. The main advantage of turning it off is that you can use std:: Stl types instead of ours. E.g. std::string vs AWS::String

**Comment by atlascoder** [comment]
Thank you, @JonathanHenson , it was actually my often question - what is the need of using Aws::String, Aws::Map, ets., instead of std::.
But nevertheless if I use custom allocation or not - I suppose that I have to use allocators from Aws:: - I mean Aws::New, Aws::Delete, Aws::MakeShared - right?

**Comment by JonathanHenson** [comment]
Nope, if you don’t use it, you can just do normal c++ things, that was the point of the design.

The reason you would use Aws:: stuff is so you don’t have to change your code if it ever gets turned on.

**Comment by atlascoder** [comment]
Great, thank you!
I've just made a factory-like class that provides for calls and and does AWSClient::DisableRequestProcessing for all created clients before Aws::ShutdownAPI - and problem is solved now.
